hashes/src/hmac.rs

```diff
@@ -7,70 +7,256 @@
 
 //! Hash-based Message Authentication Code (HMAC).
 
-use core::{convert, fmt, str};
+#[cfg(all(feature = "alloc", not(feature = "std")))]
+use alloc::vec::Vec;
+use core::{fmt, ops, str};
 
-#[cfg(feature = "serde")]
-use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use hex::DisplayHex;
+
+use crate::{sha256, sha512, FromSliceError, HashEngine};
 
-use crate::{FromSliceError, Hash, HashEngine};
+/// Returns an engine for computing `HMAC-SHA256`.
+///
+/// Equivalent to `hmac::Engine::<sha256::Engine>::new(key)`.
+pub fn new_engine_sha256(key: &[u8]) -> Engine<sha256::Engine> {
+    Engine::<sha256::Engine>::new(key)
+}
 
-/// A hash computed from a RFC 2104 HMAC. Parameterized by the underlying hash function.
+/// Returns an engine for computing `HMAC-SHA512`.
+///
+/// Equivalent to `hmac::Engine::<sha512::Engine>::new(key)`.
+pub fn new_engine_sha512(key: &[u8]) -> Engine<sha512::Engine> {
+    Engine::<sha512::Engine>::new(key)
+}
+
+/// A hash computed from a RFC 2104 HMAC. Parameterized by the size of the underlying hash function.
 #[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
-#[repr(transparent)]
-pub struct Hmac<T: Hash>(T);
+pub struct Hash<const N: usize>([u8; N]);
```

Comment:
This loses the protection against mixing incompatible HMACs. Do we really want that? And if yes, why not just remove the type entirely and use an array produced by a function?

Reply:
Good point, I also don't really like the way this turned out at the callsites, using an int is cumbersome. I'm going to play further with this.

Reply:
Please see: https://github.com/rust-bitcoin/rust-bitcoin/pull/2770#issuecomment-2141019783

Reply:
Flagging that I had another play with this and I do not have a good solution.